types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,704 @@
+	wl_resource_destroy(resource);
+}
+
+static struct wp_drm_lease_v1_interface lease_impl = {
+	.destroy = drm_lease_v1_handle_destroy,
+};
+
+static void drm_lease_request_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_drm_lease_request_v1 *request =
+		drm_lease_request_v1_from_resource(resource);
+	drm_lease_request_v1_destroy(request);
+}
+
+static void drm_lease_request_v1_handle_request_connector(
+		struct wl_client *client, struct wl_resource *request_resource,
+		struct wl_resource *connector_resource) {
+	struct wlr_drm_lease_request_v1 *request =
+		drm_lease_request_v1_from_resource(request_resource);
+	if (!request) {
+		wlr_log(WLR_ERROR, "Request has been destroyed");
+		return;
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector =
+		drm_lease_connector_v1_from_resource(connector_resource);
+
+	if (!connector) {
+		/* This connector offer has been withdrawn or is leased */
+		wlr_log(WLR_ERROR, "Failed to request connector");
+		request->invalid = true;
+		return;
+	}
+
+	wlr_log(WLR_DEBUG, "Requesting connector %s", connector->output->name);
+
+	if (request->device != connector->device) {
+		wlr_log(WLR_ERROR, "The connector belongs to another device");
+		wl_resource_post_error(request_resource,
+				WP_DRM_LEASE_REQUEST_V1_ERROR_WRONG_DEVICE,
+				"The requested connector belongs to another device");
+		return;
+	}
+
+	for (size_t i = 0; i < request->n_connectors; ++i) {
+		struct wlr_drm_lease_connector_v1 *tmp = request->connectors[i];
+
+		if (connector == tmp) {
+			wlr_log(WLR_ERROR, "The connector has already been requested");
+			wl_resource_post_error(request_resource,
+					WP_DRM_LEASE_REQUEST_V1_ERROR_DUPLICATE_CONNECTOR,
+					"The connector has already been requested");
+			return;
+		}
+	}
+
+	size_t n_connectors = request->n_connectors + 1;
+
+	struct wlr_drm_lease_connector_v1 **tmp_connectors =
+			realloc(request->connectors,
+			n_connectors * sizeof(struct wlr_drm_lease_connector_v1 *));
+	if (!tmp_connectors) {
+		wlr_log(WLR_ERROR, "Failed to grow connectors request array");
+		return;
+	}
+
+	request->connectors = tmp_connectors;
+	request->connectors[request->n_connectors] = connector;
+	request->n_connectors = n_connectors;
+}
+
+static void drm_lease_request_v1_handle_submit(
+		struct wl_client *client, struct wl_resource *resource, uint32_t id) {
+	uint32_t version = wl_resource_get_version(resource);
+	struct wl_resource *lease_resource = wl_resource_create(client,
+			&wp_drm_lease_v1_interface, version, id);
+	if (!lease_resource) {
+		wlr_log(WLR_ERROR, "Failed to allocate wl_resource");
+		wl_resource_post_no_memory(resource);
+		return;
+	}
+
+	struct wlr_drm_lease_request_v1 *request =
+			drm_lease_request_v1_from_resource(resource);
+	if (!request) {
+		wlr_log(WLR_ERROR, "Request has been destroyed");
+		wp_drm_lease_v1_send_finished(lease_resource);
+		return;
+	}
+
+	wl_resource_set_implementation(lease_resource, &lease_impl, NULL,
```

Comment:
Instead of a manual `wl_resource_set_user_data` call, we can move this `wl_resource_set_implementation` call below.